src/empathy.1
.TH EMPATHY "1" "February 2010" "Telepathy project" "User Commands"
.SH NAME
empathy \- GNOME multi-protocol chat and call client
.SH SYNOPSIS
empathy
.SH DESCRIPTION
GNOME multi-protocol chat and call client
Instant messaging program supporting text, voice, video, file transfers
and inter-application communication over many different protocols,
including: AIM, MSN, Google Talk (Jabber/XMPP), Facebook, Yahoo!, Salut,
Gadu-Gadu, Groupwise, ICQ and QQ.
.PP
The main user interface consists of a contact list window and an icon in the
notification area.
.SH OPTIONS
.TP
.BR \-n ,\  \-\-no\-connect
Don't connect on startup
.TP
.BR \-h ,\  \-\-start\-hidden
Don't display the contact list or any other dialogs on startup
.TP
.BR \-v ,\  \-\-version
Show version
.SH ENVIRONMENT
.TP
.BI EMPATHY_LOGFILE= filename
If set, debug output will go to the given file rather than to stderr.
.TP
.BI EMPATHY_DEBUG= type
May be set to "all" for full debug output, or various undocumented options
(which may change from release to release) to filter the output.
.SH AUTHOR
This manual page was written by
.MT bigon@debian.org
Laurent Bigonville
.ME ,
for the Debian GNU/Linux system (but may be used by others).
.SH SEE ALSO
.BR empathy-accounts (1),
.UR http://telepathy.freedesktop.org/
The Telepathy project
.UE ,
.UR http://live.gnome.org/Empathy
The Empathy project
.UE .


NEWS
NEW in 2.31.5.1 (13/07/2010)
===============

A small release fixing detection of libnm-glib which was accidentally
broken in 2.31.5.

Bugs fixed:
 - Fixed #624202, empathy requires libnm_glib instead of libnm-glib (Laurent Bigonville)

Translations:
 - Updated gl Translation (Fran Diéguez)
 - Updated he Translation (Yaron Shahrabani)

NEW in 2.31.5 (12/07/2010)
=============

The "Please popup chats in my face!" release.

Empathy finally gained an option to automatically present incoming events
rather than making the status icon blink.

The notifications bubbles now contain more buttons (if your notify server
supports actions) allowing you to quickly accept or decline incoming events.

The audio/video client now lives in its own process: 'empathy-av'.

Empathy can now build with GTK+ 3. If GTK+ 3 is present during configure it
will be used and Empathy will requier updated versions of other dependencies:
  - libcanberra-gtk3 ≥ 0.25
  - libnotify ≥ 0.5.1
  - libunique ≥ 2.90.1
  - WebKitGTK+ ≥ 1.3.2 (optional)
  - nautilus-sendto ≥ 2.90.0 (optional)
If GTK+ 3 is not detected or if --enable-gtk3=no is passed to configure, then
GTK+ 2 is used and the other dependencies are unchanged.


Bugs fixed:
 - Fixed #585914, Notification icon must be clicked before user can see new messages (Guillaume Desmottes)
 - Fixed #597124, Should make better use of libnotify actions (Guillaume Desmottes)
 - Fixed #599161, Audio/Video client should be a separated process (Guillaume Desmottes)
 - Fixed #617405, Empathy's short (menu) description should focus on its use, not the technology (Guillaume Desmottes)
 - Fixed #621753, Should be able to build with GTK+3 (Guillaume Desmottes)
 - Fixed #622729, duplicated messages when opening chat (Guillaume Desmottes)
 - Fixed #623054, Should re-add data/org.gnome.Empathy.gschema.xml.in to POTFILES.in (Danielle Madeley)
 - Fixed #623112, Bip backlog are not displayed any more (Guillaume Desmottes)
 - Fixed #623117, Debug window should look for TP clients (Guillaume Desmottes)
 - Fixed #623358, Use TpSimpleHandler to handle text channels (Guillaume Desmottes)
 - Fixed #623396, Fixed some leaks (Guillaume Desmottes)
 - Fixed #623679, /join broken (Guillaume Desmottes)
 - Fixed #623756, empathy-CRITICAL **: empathy_chatroom_is_always_urgent: assertion `EMPATHY_IS_CHATROOM (chatroom)' failed (Sjoerd Simons)
 - Fixed #623817, Sync man pages with Debian (Laurent Bigonville, Guillaume Desmottes)
 - Fixed #623914, since tp-logger history is shown as normal messages in adium themes (Guillaume Desmottes)
 - Fixed #623937, empathy -h does not start hidden (Guillaume Desmottes)
 - Fixed #623972, Toggling [ ] Enabled in the accounts dialog with the keyboard doesn't work (Will Thompson)
 - Fixed #624054, Fix a memory leak in EmpathyAvatar (Philip Withnall)

Translations:
 - Updated es Translation (Jorge González)
 - Updated et Translation (Ivar Smolin)
 - Updated gl Translation (Fran Diéguez)
 - Updated he Translation (Yaron Shahrabani)
 - Updated nb Translation (Kjartan Maraas)
 - Updated sl Translation (Matej Urbančič)

NEW in 2.31.4 (28/06/2010)
=============

A lot of new things in this release!

Empathy now has an awesome contact live search feature.
See http://kaserf.blogspot.com/2010/06/empathy-live-search.html for details.

Empathy doesn't log messages any more but relies on a new component,
the telepathy-logger, to do so. As a result the feature allowing you to tag
contacts as favourites is now built-in.

The contact list and status icon are now a proper Telepathy Approver. That
means external applications can now approve channels as well and Empathy will
handle them nicely.

This release makes a few steps forward to the bright GNOME 3 future by removing
deprecated GTK+ symbols and switching from GConf to GSettings.

Dependencies:

* telepathy-logger ≥ 0.1.2 (0.1.3 recommended)
* telepathy-glib ≥ 0.11.7
* libunique ≥ 0.1.2
* GTK+ ≥ 2.21.2

Bugs fixed:
 - Fixed #586476, Use accessor functions instead direct access (Danielle Madeley)
 - Fixed #588922, Empathy should use ContactInfo (Xavier Claessens)
 - Fixed #589225, We should check if the contact supports vnc stream tube for Share my desktop (Guillaume Desmottes)
 - Fixed #599158, Contact list and status icon should be an Approver (Guillaume Desmottes)
 - Fixed #608166, Add contacts live search bar (Felix Kaser, Xavier Claessens)
 - Fixed #610956, Should depend on telepathy-logger (Guillaume Desmottes)
 - Fixed #616362, Port Empathy to GSettings (Danielle Madeley)
 - Fixed #619873, search-as-you-type should only match beginnings of words (Xavier Claessens)
 - Fixed #620903, some messages get dropped when using adium themes (Guillaume Desmottes)
 - Fixed #621023, empathy does not set geoclue requirements before asking for address/position (Laurent Bigonville)
 - Fixed #621089, The live search don't always have focus in main window (Xavier Claessens)
 - Fixed #621409, empathy-2.30.1.1: tests fail (Guillaume Desmottes)
 - Fixed #621545, Stable empathy passes GTK_DISABLE_SINGLE_INCLUDES and doesn't require libunique that is fine with that (Xavier Claessens)
 - Fixed #621703, Here come a couple of UI tweaks (Will Thompson)
 - Fixed #621789, crash when closing chat with still-open notification bubble (Guillaume Desmottes)
 - Fixed #621797, Doesn't display outgoing message (Guillaume Desmottes)
 - Fixed #621894, Should port to new tp-logger API (Guillaume Desmottes)
 - Fixed #621998, Should stop using ShareWithTube() (Guillaume Desmottes)
 - Fixed #622201, Can't connect chatroom if space after adress (Guillaume Desmottes)
 - Fixed #622266, Should decide which default font to use for chat view (Xavier Claessens)
 - Fixed #622273, Crash when closing the log viewer while searching (Guillaume Desmottes)
 - Fixed #622368, Audio/Video: Use TpSimpleHandler (Guillaume Desmottes)
 - Fixed #622370, Should bold dates having logs (Guillaume Desmottes)
 - Fixed #622381, Debug dialog: display Logger logs (Guillaume Desmottes)

src/empathy.c
/*
 * Copyright (C) 2007-2009 Collabora Ltd.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301  USA
 *
 * Authors: Xavier Claessens <xclaesse@gmail.com>
 */

#include <config.h>

#include <stdlib.h>
#include <errno.h>
#include <string.h>

#include <glib.h>
#include <glib/gstdio.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <unique/unique.h>

#ifdef HAVE_LIBCHAMPLAIN
#include <clutter-gtk/clutter-gtk.h>
#endif

#include <libnotify/notify.h>

#include <telepathy-glib/account-manager.h>
#include <telepathy-glib/dbus.h>
#include <telepathy-glib/debug-sender.h>
#include <telepathy-glib/util.h>
#include <telepathy-glib/connection-manager.h>
#include <telepathy-glib/interfaces.h>

#include <telepathy-logger/log-manager.h>

#include <libempathy/empathy-idle.h>
#include <libempathy/empathy-utils.h>
#include <libempathy/empathy-chatroom-manager.h>
#include <libempathy/empathy-account-settings.h>
#include <libempathy/empathy-connectivity.h>
#include <libempathy/empathy-connection-managers.h>
#include <libempathy/empathy-dispatcher.h>
#include <libempathy/empathy-dispatch-operation.h>
#include <libempathy/empathy-ft-factory.h>
#include <libempathy/empathy-gsettings.h>
#include <libempathy/empathy-tp-chat.h>

#include <libempathy-gtk/empathy-ui-utils.h>
#include <libempathy-gtk/empathy-location-manager.h>

#include "empathy-main-window.h"
#include "empathy-accounts-common.h"
#include "empathy-accounts-dialog.h"
#include "empathy-chat-manager.h"
#include "empathy-status-icon.h"
#include "empathy-ft-manager.h"

#include "extensions/extensions.h"

#define DEBUG_FLAG EMPATHY_DEBUG_OTHER
#include <libempathy/empathy-debug.h>

static gboolean start_hidden = FALSE;
static gboolean no_connect = FALSE;

static void account_manager_ready_cb (GObject *source_object,
    GAsyncResult *result,
    gpointer user_data);

static void
dispatch_cb (EmpathyDispatcher *dispatcher,
    EmpathyDispatchOperation *operation,
    gpointer user_data)
{
  GQuark channel_type;

  channel_type = empathy_dispatch_operation_get_channel_type_id (operation);

  if (channel_type == TP_IFACE_QUARK_CHANNEL_TYPE_FILE_TRANSFER)
    {
      EmpathyFTFactory *factory;

      factory = empathy_ft_factory_dup_singleton ();

      /* if the operation is not incoming, don't claim it,
       * as it might have been triggered by another client, and
       * we are observing it.
       */
      if (empathy_dispatch_operation_is_incoming (operation))
        empathy_ft_factory_claim_channel (factory, operation);
    }
}

static void
use_conn_notify_cb (GSettings *gsettings,
    const gchar *key,
    gpointer     user_data)
{
  EmpathyConnectivity *connectivity = user_data;

  empathy_connectivity_set_use_conn (connectivity,
      g_settings_get_boolean (gsettings, key));
}

static void
migrate_config_to_xdg_dir (void)
{
  gchar *xdg_dir, *old_dir, *xdg_filename, *old_filename;
  int i;
  GFile *xdg_file, *old_file;
  static const gchar* filenames[] = {
    "geometry.ini",
    "irc-networks.xml",
    "chatrooms.xml",
    "contact-groups.xml",
    "status-presets.xml",
    "accels.txt",
    NULL
  };

  xdg_dir = g_build_filename (g_get_user_config_dir (), PACKAGE_NAME, NULL);
  if (g_file_test (xdg_dir, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_DIR))
    {
      /* xdg config dir already exists */
      g_free (xdg_dir);
      return;
    }

  old_dir = g_build_filename (g_get_home_dir (), ".gnome2",
      PACKAGE_NAME, NULL);
  if (!g_file_test (old_dir, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_DIR))
    {
      /* old config dir didn't exist */
      g_free (xdg_dir);
      g_free (old_dir);
      return;
    }

  if (g_mkdir_with_parents (xdg_dir, (S_IRUSR | S_IWUSR | S_IXUSR)) == -1)
    {
      DEBUG ("Failed to create configuration directory; aborting migration");
      g_free (xdg_dir);
      g_free (old_dir);
      return;
    }

  for (i = 0; filenames[i]; i++)
    {
      old_filename = g_build_filename (old_dir, filenames[i], NULL);
      if (!g_file_test (old_filename, G_FILE_TEST_EXISTS))
        {
          g_free (old_filename);
          continue;
        }
      xdg_filename = g_build_filename (xdg_dir, filenames[i], NULL);
      old_file = g_file_new_for_path (old_filename);
      xdg_file = g_file_new_for_path (xdg_filename);

      if (!g_file_move (old_file, xdg_file, G_FILE_COPY_NONE,
          NULL, NULL, NULL, NULL))
        DEBUG ("Failed to migrate %s", filenames[i]);

      g_free (old_filename);
      g_free (xdg_filename);
      g_object_unref (old_file);
      g_object_unref (xdg_file);
    }

  g_free (xdg_dir);
  g_free (old_dir);
}

static void
show_accounts_ui (GdkScreen *screen,
    gboolean if_needed)
{
  empathy_accounts_dialog_show_application (screen,
      NULL, if_needed, start_hidden);
}

static UniqueResponse
unique_app_message_cb (UniqueApp *unique_app,
    gint command,
    UniqueMessageData *data,
    guint timestamp,
    gpointer user_data)
{
  GtkWindow *window = user_data;
  TpAccountManager *account_manager;

  DEBUG ("Other instance launched, presenting the main window. "
      "Command=%d, timestamp %u", command, timestamp);

  /* XXX: the standalone app somewhat breaks this case, since
   * communicating it would be a pain */

  /* We're requested to show stuff again, disable the start hidden global
   * in case the accounts wizard wants to pop up.
   */
  start_hidden = FALSE;

  gtk_window_set_screen (GTK_WINDOW (window),
      unique_message_data_get_screen (data));
  gtk_window_set_startup_id (GTK_WINDOW (window),
      unique_message_data_get_startup_id (data));
  gtk_window_present_with_time (GTK_WINDOW (window), timestamp);
  gtk_window_set_skip_taskbar_hint (window, FALSE);

  account_manager = tp_account_manager_dup ();
  tp_account_manager_prepare_async (account_manager, NULL,
      account_manager_ready_cb, NULL);
  g_object_unref (account_manager);

  return UNIQUE_RESPONSE_OK;
}

static gboolean show_version_cb (const char *option_name,
    const char *value,
    gpointer data,
    GError **error) G_GNUC_NORETURN;

static gboolean
show_version_cb (const char *option_name,
    const char *value,
    gpointer data,
    GError **error)
{
  g_print ("%s\n", PACKAGE_STRING);

  exit (EXIT_SUCCESS);
}

static void
new_incoming_transfer_cb (EmpathyFTFactory *factory,
    EmpathyFTHandler *handler,
    GError *error,
    gpointer user_data)
{
  if (error)
    empathy_ft_manager_display_error (handler, error);
  else
    empathy_receive_file_with_file_chooser (handler);
}

static void
new_ft_handler_cb (EmpathyFTFactory *factory,
    EmpathyFTHandler *handler,
    GError *error,
    gpointer user_data)
{
  if (error)
    empathy_ft_manager_display_error (handler, error);
  else
    empathy_ft_manager_add_handler (handler);

  g_object_unref (handler);
}

static void
account_manager_ready_cb (GObject *source_object,
    GAsyncResult *result,
    gpointer user_data)
{
  TpAccountManager *manager = TP_ACCOUNT_MANAGER (source_object);
  GError *error = NULL;
  EmpathyIdle *idle;
  EmpathyConnectivity *connectivity;
  TpConnectionPresenceType presence;
  GSettings *gsettings = g_settings_new (EMPATHY_PREFS_SCHEMA);

  if (!tp_account_manager_prepare_finish (manager, result, &error))
    {
      DEBUG ("Failed to prepare account manager: %s", error->message);
      g_error_free (error);
      return;
    }

  /* Autoconnect */
  idle = empathy_idle_dup_singleton ();
  connectivity = empathy_connectivity_dup_singleton ();

  presence = tp_account_manager_get_most_available_presence (manager, NULL,
      NULL);

  if (g_settings_get_boolean (gsettings, EMPATHY_PREFS_AUTOCONNECT) &&
      !no_connect &&
      tp_connection_presence_type_cmp_availability
          (presence, TP_CONNECTION_PRESENCE_TYPE_OFFLINE)
            <= 0)
      /* if current state is Offline, then put it online */
      empathy_idle_set_state (idle, TP_CONNECTION_PRESENCE_TYPE_AVAILABLE);

  /* Pop up the accounts dialog if we don't have any account */
  if (!empathy_accounts_has_accounts (manager))
    show_accounts_ui (gdk_screen_get_default (), TRUE);

  g_object_unref (idle);
  g_object_unref (connectivity);
  g_object_unref (gsettings);
}

static EmpathyDispatcher *
setup_dispatcher (void)
{
  EmpathyDispatcher *d;
  GPtrArray *filters;
  struct {
    const gchar *channeltype;
    TpHandleType handletype;
  } types[] = {
    /* file transfer to contacts */
    { TP_IFACE_CHANNEL_TYPE_FILE_TRANSFER, TP_HANDLE_TYPE_CONTACT  },
  };
  GHashTable *asv;
  guint i;

  /* Setup the an extended Client.Handler that matches everything we can do */
  filters = g_ptr_array_new ();
  for (i = 0 ; i < G_N_ELEMENTS (types); i++)
    {
      asv = tp_asv_new (
        TP_IFACE_CHANNEL ".ChannelType", G_TYPE_STRING, types[i].channeltype,
        TP_IFACE_CHANNEL ".TargetHandleType", G_TYPE_INT, types[i].handletype,
        NULL);

      g_ptr_array_add (filters, asv);
    }

  d = empathy_dispatcher_new (PACKAGE_NAME"MoreThanMeetsTheEye", filters, NULL);

  g_ptr_array_foreach (filters, (GFunc) g_hash_table_destroy, NULL);
  g_ptr_array_free (filters, TRUE);

  return d;
}

static void
account_status_changed_cb (TpAccount *account,
    guint old_status,
    guint new_status,
    guint reason,
    gchar *dbus_error_name,
    GHashTable *details,
    EmpathyChatroom *room)
{
  TpConnection *conn;

  conn = tp_account_get_connection (account);
  if (conn == NULL)
    return;

  empathy_dispatcher_join_muc (conn,
      empathy_chatroom_get_room (room), EMPATHY_DISPATCHER_NON_USER_ACTION);
}

static void
account_manager_chatroom_ready_cb (GObject *source_object,
    GAsyncResult *result,
    gpointer user_data)
{
  TpAccountManager *account_manager = TP_ACCOUNT_MANAGER (source_object);
  EmpathyChatroomManager *chatroom_manager = user_data;
  GList *accounts, *l;
  GError *error = NULL;

  if (!tp_account_manager_prepare_finish (account_manager, result, &error))
    {
      DEBUG ("Failed to prepare account manager: %s", error->message);
      g_error_free (error);
      return;
    }

  accounts = tp_account_manager_get_valid_accounts (account_manager);

  for (l = accounts; l != NULL; l = g_list_next (l))
    {
      TpAccount *account = TP_ACCOUNT (l->data);
      TpConnection *conn;
      GList *chatrooms, *p;

      conn = tp_account_get_connection (account);

      chatrooms = empathy_chatroom_manager_get_chatrooms (
          chatroom_manager, account);

      for (p = chatrooms; p != NULL; p = p->next)
        {
          EmpathyChatroom *room = EMPATHY_CHATROOM (p->data);

          if (!empathy_chatroom_get_auto_connect (room))
            continue;

          if (conn == NULL)
            {
              g_signal_connect (G_OBJECT (account), "status-changed",
                  G_CALLBACK (account_status_changed_cb), room);
            }
          else
            {
              empathy_dispatcher_join_muc (conn,
                  empathy_chatroom_get_room (room),
                  EMPATHY_DISPATCHER_NON_USER_ACTION);
            }
        }

      g_list_free (chatrooms);
    }

  g_list_free (accounts);
}

static void
chatroom_manager_ready_cb (EmpathyChatroomManager *chatroom_manager,
    GParamSpec *pspec,
    gpointer user_data)
{
  TpAccountManager *account_manager = user_data;

  tp_account_manager_prepare_async (account_manager, NULL,
      account_manager_chatroom_ready_cb, chatroom_manager);
}

static void
empathy_idle_set_auto_away_cb (GSettings *gsettings,
				const gchar *key,
				gpointer user_data)
{
	EmpathyIdle *idle = user_data;

	empathy_idle_set_auto_away (idle,
      g_settings_get_boolean (gsettings, key));
}

int
main (int argc, char *argv[])
{
#ifdef HAVE_GEOCLUE
  EmpathyLocationManager *location_manager = NULL;
#endif
  EmpathyStatusIcon *icon;
  EmpathyDispatcher *dispatcher;
  TpAccountManager *account_manager;
  TplLogManager *log_manager;
  EmpathyChatroomManager *chatroom_manager;
  EmpathyFTFactory  *ft_factory;
  GtkWidget *window;
  EmpathyIdle *idle;
  EmpathyConnectivity *connectivity;
  EmpathyChatManager *chat_manager;
  GError *error = NULL;
  UniqueApp *unique_app;
  gboolean chatroom_manager_ready;
  gboolean autoaway = TRUE;
#ifdef ENABLE_DEBUG
  TpDebugSender *debug_sender;
#endif
  GSettings *gsettings;

  GOptionContext *optcontext;
  GOptionEntry options[] = {
      { "no-connect", 'n',
        0, G_OPTION_ARG_NONE, &no_connect,
        N_("Don't connect on startup"),
        NULL },
      { "start-hidden", 'h',
        0, G_OPTION_ARG_NONE, &start_hidden,
        N_("Don't display the contact list or any other dialogs on startup"),
        NULL },
      { "version", 'v',
        G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, show_version_cb,
        NULL, NULL },
      { NULL }
  };

  /* Init */
  g_thread_init (NULL);
  empathy_init ();

  optcontext = g_option_context_new (N_("- Empathy IM Client"));
  g_option_context_add_group (optcontext, gtk_get_option_group (TRUE));
#ifdef HAVE_LIBCHAMPLAIN
  g_option_context_add_group (optcontext, clutter_get_option_group ());
#endif
  g_option_context_add_main_entries (optcontext, options, GETTEXT_PACKAGE);

  if (!g_option_context_parse (optcontext, &argc, &argv, &error)) {
    g_print ("%s\nRun '%s --help' to see a full list of available command line options.\n",
        error->message, argv[0]);
    g_warning ("Error in empathy init: %s", error->message);
    return EXIT_FAILURE;
  }

  g_option_context_free (optcontext);

  empathy_gtk_init ();
  g_set_application_name (_(PACKAGE_NAME));

  gtk_window_set_default_icon_name ("empathy");
  textdomain (GETTEXT_PACKAGE);

#ifdef ENABLE_DEBUG
  /* Set up debug sender */
  debug_sender = tp_debug_sender_dup ();
  g_log_set_default_handler (tp_debug_sender_log_handler, G_LOG_DOMAIN);
#endif

  unique_app = unique_app_new ("org.gnome."PACKAGE_NAME, NULL);

  if (unique_app_is_running (unique_app))
    {
      unique_app_send_message (unique_app, UNIQUE_ACTIVATE, NULL);

      g_object_unref (unique_app);
      return EXIT_SUCCESS;
    }

  notify_init (_(PACKAGE_NAME));

  /* Setting up Idle */
  idle = empathy_idle_dup_singleton ();

  gsettings = g_settings_new (EMPATHY_PREFS_SCHEMA);
  autoaway = g_settings_get_boolean (gsettings, EMPATHY_PREFS_AUTOAWAY);

  g_signal_connect (gsettings,
      "changed::" EMPATHY_PREFS_AUTOAWAY,
      G_CALLBACK (empathy_idle_set_auto_away_cb), idle);

  empathy_idle_set_auto_away (idle, autoaway);

  /* Setting up Connectivity */
  connectivity = empathy_connectivity_dup_singleton ();
  use_conn_notify_cb (gsettings, EMPATHY_PREFS_USE_CONN,
      connectivity);
  g_signal_connect (gsettings,
      "changed::" EMPATHY_PREFS_USE_CONN,
      G_CALLBACK (use_conn_notify_cb), connectivity);

  /* account management */
  account_manager = tp_account_manager_dup ();
  tp_account_manager_prepare_async (account_manager, NULL,
      account_manager_ready_cb, NULL);

  /* Handle channels */
  dispatcher = setup_dispatcher ();
  g_signal_connect (dispatcher, "dispatch", G_CALLBACK (dispatch_cb), NULL);

  migrate_config_to_xdg_dir ();

  /* Setting up UI */
  window = empathy_main_window_dup ();
  gtk_widget_show (window);
  icon = empathy_status_icon_new (GTK_WINDOW (window), start_hidden);

  /* Chat manager */
  chat_manager = empathy_chat_manager_dup_singleton ();

  g_signal_connect (unique_app, "message-received",
      G_CALLBACK (unique_app_message_cb), window);

  /* Logging */
  log_manager = tpl_log_manager_dup_singleton ();

  chatroom_manager = empathy_chatroom_manager_dup_singleton (NULL);

  g_object_get (chatroom_manager, "ready", &chatroom_manager_ready, NULL);
  if (!chatroom_manager_ready)
    {
      g_signal_connect (G_OBJECT (chatroom_manager), "notify::ready",
          G_CALLBACK (chatroom_manager_ready_cb), account_manager);
    }
  else
    {
      chatroom_manager_ready_cb (chatroom_manager, NULL, account_manager);
    }

  /* Create the FT factory */
  ft_factory = empathy_ft_factory_dup_singleton ();
  g_signal_connect (ft_factory, "new-ft-handler",
      G_CALLBACK (new_ft_handler_cb), NULL);
  g_signal_connect (ft_factory, "new-incoming-transfer",
      G_CALLBACK (new_incoming_transfer_cb), NULL);

  /* Location mananger */
#ifdef HAVE_GEOCLUE
  location_manager = empathy_location_manager_dup_singleton ();
#endif

  gtk_main ();

  empathy_idle_set_state (idle, TP_CONNECTION_PRESENCE_TYPE_OFFLINE);

#ifdef ENABLE_DEBUG
  g_object_unref (debug_sender);
#endif

  g_object_unref (chat_manager);
  g_object_unref (idle);
  g_object_unref (connectivity);
  g_object_unref (icon);
  g_object_unref (account_manager);
  g_object_unref (log_manager);
  g_object_unref (dispatcher);
  g_object_unref (chatroom_manager);
#ifdef HAVE_GEOCLUE
  g_object_unref (location_manager);
#endif
  g_object_unref (ft_factory);
  g_object_unref (unique_app);
  g_object_unref (gsettings);
  gtk_widget_destroy (window);

  notify_uninit ();
  xmlCleanupParser ();

  return EXIT_SUCCESS;
}
